src/client/common/platform/fileSystem.ts

```diff
@@ -177,6 +189,89 @@ export class FileSystem implements IFileSystem {
         return deferred.promise;
     }
 
+    public chmod(filePath: string, mode: string | number): Promise<void> {
+        return new Promise<void>((resolve, reject) => {
+            fileSystem.chmod(filePath, mode, (err: NodeJS.ErrnoException | null) => {
+                if (err) {
+                    return reject(err);
+                }
+                resolve();
+            });
+        });
+    }
+
+    public async move(src: string, tgt: string) {
+        await fs.rename(src, tgt);
+    }
+
+    public createReadStream(filePath: string): fileSystem.ReadStream {
+        return fileSystem.createReadStream(filePath);
+    }
+
+    public createWriteStream(filePath: string): fileSystem.WriteStream {
+        return fileSystem.createWriteStream(filePath);
+    }
+
+    //=================================
+    // utils
+
+    public objectExists(filePath: string, statCheck: (s: fs.Stats) => boolean): Promise<boolean> {
+        return new Promise<boolean>(resolve => {
+            // Note that we are using stat() rather than lstat().  This
+            // means that any symlinks are getting resolved.
+            fs.stat(filePath, (error, stats) => {
+                if (error) {
+                    return resolve(false);
+                }
+                return resolve(statCheck(stats));
+            });
+        });
+    }
+    public fileExists(filePath: string): Promise<boolean> {
+        return this.objectExists(filePath, stats => stats.isFile());
+    }
+    public fileExistsSync(filePath: string): boolean {
+        return fs.existsSync(filePath);
+    }
+    public directoryExists(filePath: string): Promise<boolean> {
+        return this.objectExists(filePath, stats => stats.isDirectory());
+    }
+
+    public async getSubDirectories(dirname: string): Promise<string[]> {
+        let files: [string, FileType][];
+        try {
+            files = await this.listdir(dirname);
+        } catch {
+            // We're only preserving pre-existng behavior here...
+            return [];
+        }
+        return files
+            .filter(([_file, fileType]) => {
+                // We preserve the pre-existing behavior of following
+                // symlinks.
+                return (fileType & FileType.Directory) > 0;
+            })
+            .map(([filename, _ft]) => filename);
+    }
+    public async getFiles(dirname: string): Promise<string[]> {
+        let files: [string, FileType][];
+        try {
+            files = await this.listdir(dirname);
+        } catch (err) {
+            if (!await fs.pathExists(dirname)) {
+                return [];
+            }
+            throw err; // re-throw
+        }
+        return files
+            .filter(([_file, fileType]) => {
+                // We preserve the pre-existing behavior of following
+                // symlinks.
+                return (fileType & FileType.Directory) === 0;
```

Comment:
What happens if we have something as `unknown`. Wouldn't that get included into this, when it shouldn't!
I.e. assume files contains, `file, directory, file + symlink, unknown`, now the filtered list contains `file, file + symlink, unknown`

Reply:
Hmm, yeah, I need to be more careful in matching the existing behavior:

```suggestion
                return (fileType | FileType.File) > 0;
```